src/resolve.ts

```diff
@@ -0,0 +1,47 @@
+import {NonspatialScaleChannel, SCALE_CHANNELS, SPATIAL_SCALE_CHANNELS, SpatialScaleChannel, UNIT_SCALE_CHANNELS} from './channel';
+import * as log from './log';
+import {contains} from './util';
+
+export type ResolveMode = 'independent' | 'shared';
+
+export interface SpatialResolve {
+  scale: ResolveMode;
+  axis?: ResolveMode;
+}
+
+export interface NonspatialResolve {
+  scale: ResolveMode;
+  legend?: ResolveMode;
+}
+
+export type Resolve = SpatialResolve | NonspatialResolve;
+
+export function isSpatialResolve(resolve: Resolve): resolve is SpatialResolve {
+  return 'axis' in resolve;
+}
+
+export function isNonspatialResolve(resolve: Resolve): resolve is NonspatialResolve {
+  return 'legend' in resolve;
+}
+
+export type ResolveMapping = {[P in NonspatialScaleChannel]?: NonspatialResolve} | {[P in SpatialScaleChannel]?: SpatialResolve};
+
+export function initLayerResolve(resolve: ResolveMapping): ResolveMapping {
+  UNIT_SCALE_CHANNELS.forEach(channel => {
```

Comment:
I changed the method so that it doesn't cause side effects. Having all resolves makes things much easier downstream but I agree that it is a bit wasteful.